SPI: Do not disable SPI peripheral while switching DMA buffers

LuxLuthor opened the pull request:
Removed SPI peripheral disabling when switching DMA to another buffer. When using hardware chip select this would cause to stop driving CS pin when swhitching buffers. This is different (and wrong) than when used software CS.

Fixes #28833

Signed-off-by: Łukasz Mazur <lukasz.mazur@hidglobal.com>

Hid665 committed:
SPI: Do not disable SPI peripheral while switching DMA buffers

Removed SPI peripheral disabling when switching DMA to another buffer.
When using hardware chip select this would cause to stop driving
CS pin when swhitching buffers. This is different (and wrong) than
when used software CS.

Fixes #28833

Signed-off-by: Łukasz Mazur <lukasz.mazur@hidglobal.com>

tbursztyka commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -729,8 +729,7 @@ static int transceive_dma(const struct device *dev,
 		while (LL_SPI_IsActiveFlag_BSY(spi) == 1) {
 		}
-
-		LL_SPI_Disable(spi);
```
Btw, aren't lines 689 to 691 superfluous? (At this stage, cs nor dma are enabled: the transaction has yet to start)

LuxLuthor replied:
Probably yes but this PR solves only HW CS problem. I don't know what was the idea of the original author.

lowlander replied:
I think that code ended up there after quit a bit of conversation in the PR that changed that code, I guess nobody had a use case/hardware that used HW CS :-/ I will test (but I won't have time until in the weekend) it on the boards I have, and if those all work I guess it should be OK.

lowlander replied:
@tbursztyka a quick test shows that line 689 and 690 can be removed, but line 691 `LL_SPI_Disable(spi);` should be changed to `LL_SPI_Enable(spi);` and than line 713 can also be removed.

tbursztyka replied:
nice, can you make a pr?

lowlander replied:
@LuxLuthor could you add those changes mentioned by me to this PR, or do you want me to make a new combined PR with those changes and your fix ?

lowlander replied:
@erwango the question was if cleanup the extra enable call in this PR, or if we make two "mini" PR's

erwango replied:
@lowlander @LuxLuthor, as long as this is done in 2 commits that are properly documented and current PR description is updated, this can be done here.

lowlander replied:
@erwango the fix is to remove the disable from the loop, so the enable in the loop makes no sense anymore, and can only be done one time before the loop. so those two changes belong in one (this) commit ? The removing of the dmadisable in front of the loop could be the second commit ?

LuxLuthor replied:
@erwango @lowlander, Yep, I will add these changes in this PR.

lowlander replied:
@erwango On the other hand maybe just merge this commit and make a cleanup PR, so @Luxluthor doesn't have to change this PR/commit ? I can make a new PR with the cleanup.

lowlander replied:
@erwango @tbursztyka should @luxluthor make one commit with removing the SPI disable/enable out of the loop and one removing the two dmadisables in front of the loop ?

tbursztyka reviewed (approved):
Just reformat your commit message so it fits whithin 72chars per line (something like that)

lowlander commented:
@erwango Since this will need some actual testing on different boards (at least one without dmamux and one with) I will not be able to seriously review this until in the weekend.

erwango commented:
> @erwango Since this will need some actual testing on different boards (at least one without dmamux and one with) I will not be able to seriously review this until in the weekend.

This is totally fine, thanks for your help in reviewing this!

lowlander commented:
Tested with nucleo_L496zg, nucleo_l053r8, nucleo_f767zi, and l4r5zi (some boards need some not yet merged PRs) and all seem to still work fine with GPIO-CS. Tested the L4R5ZI with HW CS and it looked OK.

This is the CS signal without the fix.

![spi_without_fix](https://user-images.githubusercontent.com/1266398/95652895-b03cba80-0af4-11eb-972e-775688aaa1c1.png)

And this with the fix.

![spi_with_fix](https://user-images.githubusercontent.com/1266398/95652906-c0ed3080-0af4-11eb-8cb7-eee69b83df57.png)

erwango commented:
> "lowlander self-requested a review 1 hour ago"

@lowlander, having some doubts ?

lowlander commented:
> > "lowlander self-requested a review 1 hour ago"
> 
> @lowlander, having some doubts ?

@erwango not about the fix, but if we want to do the little cleanup in this commit also, see the discussion in the code above.

Hid665 committed:
SPI: Refactoring in SPI with DMA (peripheral switching on/off)

Moved enabling SPI peripheral in front of the buffers loop.
Removed SPI DMA switching on in front of the buffers loop.

Signed-off-by: Łukasz Mazur <lukasz.mazur@hidglobal.com>

gmarull reviewed (approved):
Haven't tested myself on HW, but changes LGTM. @lowlander thanks for the signal captures!
